runtime/observability/telemetry-provider-micrometer/jvm/src/aws/smithy/kotlin/runtime/telemetry/micrometer/MicrometerMeterProvider.kt

```diff
@@ -0,0 +1,211 @@
+        MicrometerLongHistogram(
+            MicrometerDoubleHistogram(
+                meterMetadata = MeterMetadata(name, units, description),
+                meterRegistry = meterRegistry,
+            )
+        )
+
+    override fun createDoubleHistogram(name: String, units: String?, description: String?): DoubleHistogram =
+        MicrometerDoubleHistogram(
+            meterMetadata = MeterMetadata(name, units, description),
+            meterRegistry = meterRegistry,
+        )
+
+    override fun createLongGauge(
+        name: String,
+        callback: LongGaugeCallback,
+        units: String?,
+        description: String?,
+    ): AsyncMeasurementHandle = MicrometerLongGauge(
+        callback = callback,
+        meterMetadata = MeterMetadata(name, units, description),
+        meterRegistry = meterRegistry,
+    )
+
+    override fun createDoubleGauge(
+        name: String,
+        callback: DoubleGaugeCallback,
+        units: String?,
+        description: String?,
+    ): AsyncMeasurementHandle = MicrometerDoubleGauge(
+        callback = callback,
+        meterMetadata = MeterMetadata(name, units, description),
+        meterRegistry = meterRegistry,
+    )
+
+}
+
+private data class MeterMetadata(val meterName: String, val units: String?, val description: String?)
+
+private class MicrometerUpDownCounter(
+    private val meterMetadata: MeterMetadata,
+    private val meterRegistry: MeterRegistry,
+) : UpDownCounter {
+    override fun add(value: Long, attributes: Attributes, context: Context?) {
+        meterMetadata
+            .counter()
+            .tags(attributes.toTags())
+            .register(meterRegistry)
+            .increment(value.toDouble())
+    }
+}
+
+private class MicrometerMonotonicCounter(
+    private val meterMetadata: MeterMetadata,
+    private val meterRegistry: MeterRegistry,
+) : MonotonicCounter {
+    override fun add(value: Long, attributes: Attributes, context: Context?) {
+        meterMetadata
+            .counter()
+            .tags(attributes.toTags())
+            .register(meterRegistry)
+            .increment(value.toDouble())
+    }
+}
+
+private class MicrometerDoubleHistogram(
+    private val meterMetadata: MeterMetadata,
+    private val meterRegistry: MeterRegistry,
+) : DoubleHistogram {
+    override fun record(value: Double, attributes: Attributes, context: Context?) {
+        DistributionSummary.builder(meterMetadata.meterName)
+            .baseUnit(meterMetadata.units)
+            .description(meterMetadata.description)
+            .tags(attributes.toTags())
+            .publishPercentileHistogram()
+            .register(meterRegistry)
+    }
+}
+
+private class MicrometerLongHistogram(
+    private val doubleHistogram: MicrometerDoubleHistogram,
+) : LongHistogram {
+    override fun record(value: Long, attributes: Attributes, context: Context?) {
+        doubleHistogram.record(value.toDouble(), attributes, context)
+    }
+}
```

Comment:
That wouldn't work, since both `LongHistogram` and `DoubleHistogram` are essentially the same interface with different generic:
![image](https://github.com/smithy-lang/smithy-kotlin/assets/5037832/c3b4a39e-db58-4caf-b56a-0a2531fc8229)